### ISSUE TITLE
[DatePicker]: Input mask isn't set back

### ISSUE THREAD

**Issue by EricSch** [telerik/kendo-angular#807]
* Bug report

## Current behavior
I enter manually a date into the DatePicker. In the valueChange event, I check for a valid date. When  the date is invalid, I set the the DatePicker to undefined. The first time it works, the mask is set back, but the second time the mask contains still the undesired values.

## Expected behavior
When the date is invalid, the DatePicker should be reseted and the input mask again "month/day/year".

I understand, that the DatePicker value hasn't changed, so probably nothing will be triggered. But at least it should have always the same behavior or a method, to set it back manually

## Minimal reproduction of the problem with instructions
[plnkr.co/edit/QYb5A8VkHogkFA63uO5G?p=preview](plnkr.co/edit/QYb5A8VkHogkFA63uO5G?p=preview)

1. Enter manually 1/1/1
2. Control is set back to 'month/day/year'
3. Enter manually 1/2/2
4. Control isn't set back, still 1/2/2

## Environment
Latest kendo version, Angular 4 and 5

Browser:
- Chrome (desktop) version 59.0

**Comment by ggkrustev** [comment]
@EricSch, good catch! Will update the thread once the issue is fixed.

**Comment by ggkrustev** [comment]
Hi @EricSch,

Please excuse me for the late follow up. I further investigated the case and noticed smth that I omitted the first time. The component doesn't reset its input, because there is no `value` input change. Basically, this is what happens:
- `1/1/1` is entered
- `valueChanged` is triggered
-  `this.value` property is set to `undefined`
- input value is set to `month/day/year`
- `1/2/2` is typed
- `valueChanged` is raised again
- `this.value` is set to `undefined`, but it is already `undefined`
- **no change detection of the DatePicker is initiated**

This is expected and the main Angular change detection feature - *avoiding unnecessary child component updates if there are no changes*.

One feasible solution is to update `this.value` property to a valid date and then reset it back again to `undefined` in order to spin children change detection. Check the updated plunker:

http://plnkr.co/edit/QHMkEtOKclwWOYSfhtjZ?p=preview

I will close this issue, as it is not a bug in the DatePicker component.

**Comment by EricSch** [comment]
I posted this as bug because of the behavior of the control, not because of an angular behavior. I'm understand that no change event is triggered from angular side, but your control should behave consistently. I shouldn't have to do workarounds just that the control behaves how it should. **I think you are going the easy way.** 

Why the control doesn't do that? Somewhere you have to parse the values. Then the controls know if it's valid or not and the control could trigger the change.

**Comment by ggkrustev** [comment]
Hi @EricSch,

our decision is not based on possible "easy/hard way to go", but rather on the core principals of the Angular framework.

I can assure you that we investigated the case in great details (observations shared in previous reply) and we find out that *the DatePicker component will not be notified when the parent `value` field is set to the same value*. This is how the [Angular change detection](https://vsavkin.com/change-detection-in-angular-2-4f216b855d4c) is supposed to work. In other words, this is what happen
- component internal state is set to the chosen date and change event is triggered
- parent component handles change event and sets its `value` property to the same `undefined` value

After the second step,
- the DatePicker value is equal to the shown in the input
- parent component value is still `undefined`
- **no notification signal for the change (because there isn't any) is send from parent -> child through `[value] binding`**

Basically, **the DatePicker OnChanges hook is never triggered**. We cannot clear component value after `valueChange` event based on our guts feeling. The component internal state is updated only when *@Input `value` property* is updated from the outside. When there is no change, component internal state will stay untouched.

Based on the given observations and considering how Angular is indented to work in this case, I can conclude that **the reported behavior is actually expected and the DatePicker is consistent and follows the Angular core principle for change detection**.  I would suggest using the aforementioned workaround that enforces Angular change detection and ensures that DatePicker component state is always updated correctly.

**Comment by EricSch** [comment]
Hi @ggkrustev 
Thank you, I appreciate that you take time to answer.

Like I wrote, I don't talk about how Angular works. I see that, and I work daily with the core principle of Angular. I develop application for customers, they expect a certain behavior of the app, and it's my duty, to provide that. When I would always say, but that's one core principle of the Angular framework, we would be out of business. So I look for ways to give them the expected behavior.

**In my opinion, Kendo UI for Angular have to work consistently**. Same input, same output, and each time. See it as a "Core principle of the Kendo UI for Angular framework". How do I explain to our customers, that this is a Core principle of Angular, so the controls behave differently? 

**My suggestion was that the Kendo UI for Angular implements the workaround, and not we as users.** Look at your function `KendoDate.prototype.parsePart`, there is plenty of possibilities to force this.value change detection when the old value was already `null `or `undefined`. You posted the workaround here :smirk:

This is my view as customers of Kendo UI for Angular

**Comment by ggkrustev** [comment]
Hi @EricSch,

we'd love to implement the workaround if we know when the internal state of the component should be changed. Could you elaborate on how do you think this should be done? The parent component sets its `value` property, but DatePicker's OnChange hook is never called. How the DatePicker will know that the value should be reset? Shouldn't the `@Input() property` be the single source of truth in this case?

Any insights or possible ideas on how this should be handled are welcomed.

**Comment by EricSch** [comment]
Sorry, when I'm overreaching here, lot I'm assuming.

Shouldn't it concern only DateInput as the values are entered manually? There you have the HostListener 'handleInput' which will be called when the input changes, I assume. Adding a logic to force an update of the element content (with `updateElementValue` function) when old and new parsed value are null?

**Comment by ggkrustev** [comment]
Indeed, when a new view is typed only DateInput is active. Nevertheless, both `1/1/1` and `1/2/2/` are valid values for the component. It doesn't know whether one of this will be invalid based on some business condition/requirement. It can recognize only a valid date (`1/1/1`) or invalid date (`month/1/1`).

Although, we could add logic that resets the internal selected value, when we will need to know that `1/1/1` is valid or not? This clearly is controlled from outside world using `[value]` binding, which in turns is controlled by the Angular change detection mechanism.

I am still not sure how the component internal can be notified from outside that `1/1/1` is valid or not without using the Angular binding mechanism. Do you have any insights on that?

**Comment by EricSch** [comment]
I thought on a explicit method call, like reset()? Of course this would work without any binding and controlled by the business logic.

I had another thought, perhaps I could solve my issue with a dynamic min and max binding? :bowtie:

**Comment by ggkrustev** [comment]
Our thorough investigation (conducted before posting my previous reply) showed that all parent-child related components should use a-la `reset` method (or just utilize the public `writeValue`). In all other cases, the missing change, according Angular change detection system, keeps the input unchanged.

Adding ad-hoc method that makes the changes forcefully ignoring Angular could be a useful tool, but rather is out-of-scope for now. Would it be possible to post a request in our UserVoice portal. Thus we will be able to gather community feedback and monitor the interest in such component feature.

Anything that triggers change detection cycle will update the component internals. Either changing the instance of the selected value or other component properties. I suppose that min/max changes will do the trick.
